File: Code/Lesson_8_code/hash_speed.py
"""What a leaked password table is worth to the person who steals it.

    python hash_speed.py

The program does four things:

    1. It shows that two users with the same password get the same SHA-256
       row, and different bcrypt rows.
    2. It cracks the SHA-256 table with a word list, and reports the time.
    3. It attacks the bcrypt table with the same word list, and reports the
       time.
    4. It measures both functions and prints the guess rate of each one.

Every number that this program prints comes from your machine. Run it before
you read the lesson section that quotes it.
"""

import hashlib
import pathlib
import time

import bcrypt

from config import settings

WORDLIST = pathlib.Path(__file__).parent / "wordlist.txt"

# Five accounts. Two of them chose the same password, which is normal.
ACCOUNTS = [
    ("ada@example.com", "hunter2"),
    ("alan@example.com", "letmein"),
    ("grace@example.com", "hunter2"),
    ("edsger@example.com", "correcthorse"),
    ("barbara@example.com", "trustno1"),
]


def sha256_hex(password: str) -> str:
    return hashlib.sha256(password.encode()).hexdigest()


def rule(title: str) -> None:
    print(f"\n=== {title} " + "=" * max(0, 62 - len(title)))


def main() -> None:
    words = WORDLIST.read_text().split()
    print(f"word list: {len(words)} passwords")
    print(f"bcrypt rounds: {settings.bcrypt_rounds}")

    # ---------------------------------------------------------------- 1 ---
    rule("The leak, as SHA-256")
    sha_table = [(email, sha256_hex(pw)) for email, pw in ACCOUNTS]
    for email, digest in sha_table:
        print(f"{email:22} {digest}")
    print("\nAda and Grace never told each other their password.")
    print("The table tells you they share one.")

    rule("The same leak, as bcrypt")
    bcrypt_table = [
        (email, bcrypt.hashpw(pw.encode(), bcrypt.gensalt(settings.bcrypt_rounds)).decode())
        for email, pw in ACCOUNTS
    ]
    for email, digest in bcrypt_table:
        print(f"{email:22} {digest}")
    print("\nThe same two passwords now look unrelated. The salt did that.")

    # ---------------------------------------------------------------- 2 ---
    rule("Crack the SHA-256 table")
    start = time.perf_counter()
    index = {sha256_hex(w): w for w in words}
    cracked = [(email, index[d]) for email, d in sha_table if d in index]
    sha_seconds = time.perf_counter() - start
    for email, password in cracked:
        print(f"{email:22} {password}")
    print(f"\n{len(cracked)} of {len(ACCOUNTS)} accounts in {sha_seconds * 1000:.2f} ms")

    # ---------------------------------------------------------------- 3 ---
    rule("Attack the bcrypt table with the same word list")
    start = time.perf_counter()
    found = []
    for email, digest in bcrypt_table:
        for word in words:
            if bcrypt.checkpw(word.encode(), digest.encode()):
                found.append((email, word))
                break
    bcrypt_seconds = time.perf_counter() - start
    for email, password in found:
        print(f"{email:22} {password}")
    print(f"\n{len(found)} of {len(ACCOUNTS)} accounts in {bcrypt_seconds:.1f} s")
    print(f"slower by {bcrypt_seconds / sha_seconds:,.0f}x for the same result")
    print("Note the result. bcrypt does not save a password that is on the list.")

    # ---------------------------------------------------------------- 4 ---
    rule("Guess rate of each function")
    sample = b"correct horse battery staple"

    start = time.perf_counter()
    for _ in range(200_000):
        hashlib.sha256(sample).hexdigest()
    sha_rate = 200_000 / (time.perf_counter() - start)

    salt = bcrypt.gensalt(settings.bcrypt_rounds)
    start = time.perf_counter()
    for _ in range(10):
        bcrypt.hashpw(sample, salt)
    bcrypt_rate = 10 / (time.perf_counter() - start)

    space = 26 ** 8  # every lower-case password of eight letters
    print(f"SHA-256 : {sha_rate:15,.0f} guesses/second, one CPU core")
    print(f"bcrypt  : {bcrypt_rate:15,.0f} guesses/second, one CPU core")
    print(f"ratio   : {sha_rate / bcrypt_rate:15,.0f}x")
    print(f"\nA password of eight lower-case letters: {space:,} possibilities.")
    print(f"SHA-256 : {space / sha_rate / 3600:,.1f} hours on this one core")
    print(f"bcrypt  : {space / bcrypt_rate / 31_557_600:,.0f} years on this one core")
    print("\nA real attacker owns a GPU farm, not one core of a laptop.")
    print("The ratio is what matters, and the ratio does not change.")


if __name__ == "__main__":
    main()

File: Code/js/Lesson_11_code/src/hashSpeed.ts
/**
 * The TypeScript twin of `Code/Lesson_8_code/hash_speed.py`.
 *
 *     node src/hashSpeed.ts
 *
 * Same four steps, same word list, same conclusion. Compare the guess rates
 * with the Python run. The language does not change the answer, because
 * bcrypt is the same C code in both.
 */

import bcrypt from 'bcrypt';
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

import { config } from './config.ts';

const WORDLIST = path.join(import.meta.dirname, '..', 'wordlist.txt');

const ACCOUNTS: ReadonlyArray<readonly [string, string]> = [
  ['ada@example.com', 'hunter2'],
  ['alan@example.com', 'letmein'],
  ['grace@example.com', 'hunter2'],
  ['edsger@example.com', 'correcthorse'],
  ['barbara@example.com', 'trustno1'],
];

function rule(title: string): void {
  console.log(`\n=== ${title} ${'='.repeat(Math.max(0, 62 - title.length))}`);
}

const words = fs.readFileSync(WORDLIST, 'utf8').split('\n').filter(Boolean);
console.log(`word list: ${words.length} passwords`);
console.log(`bcrypt rounds: ${config.bcryptRounds}`);

// ------------------------------------------------------------------ 1 ---
rule('The leak, as SHA-256');
const shaTable = ACCOUNTS.map(
  ([email, password]) =>
    [email, crypto.createHash('sha256').update(password).digest('hex')] as const,
);
for (const [email, digest] of shaTable) console.log(email.padEnd(22), digest);
console.log('\nAda and Grace never told each other their password.');
console.log('The table tells you they share one.');

rule('The same leak, as bcrypt');
const bcryptTable = ACCOUNTS.map(
  ([email, password]) => [email, bcrypt.hashSync(password, config.bcryptRounds)] as const,
);
for (const [email, digest] of bcryptTable) console.log(email.padEnd(22), digest);
console.log('\nThe same two passwords now look unrelated. The salt did that.');

// ------------------------------------------------------------------ 2 ---
rule('Crack the SHA-256 table');
let start = performance.now();
const index = new Map(
  words.map((word) => [crypto.createHash('sha256').update(word).digest('hex'), word]),
);
const cracked = shaTable
  .map(([email, digest]) => [email, index.get(digest)] as const)
  .filter((pair): pair is readonly [string, string] => pair[1] !== undefined);
const shaMs = performance.now() - start;
for (const [email, password] of cracked) console.log(email.padEnd(22), password);
console.log(`\n${cracked.length} of ${ACCOUNTS.length} accounts in ${shaMs.toFixed(2)} ms`);

// ------------------------------------------------------------------ 3 ---
rule('Attack the bcrypt table with the same word list');
start = performance.now();
const found: Array<readonly [string, string]> = [];
for (const [email, digest] of bcryptTable) {
  for (const word of words) {
    if (bcrypt.compareSync(word, digest)) {
      found.push([email, word]);
      break;
    }
  }
}
const bcryptSeconds = (performance.now() - start) / 1000;
for (const [email, password] of found) console.log(email.padEnd(22), password);
console.log(`\n${found.length} of ${ACCOUNTS.length} accounts in ${bcryptSeconds.toFixed(1)} s`);
console.log(
  `slower by ${Math.round((bcryptSeconds * 1000) / shaMs).toLocaleString('en-US')}x for the same result`,
);
console.log('Note the result. bcrypt does not save a password that is on the list.');

// ------------------------------------------------------------------ 4 ---
rule('Guess rate of each function');
const sample = 'correct horse battery staple';

start = performance.now();
for (let i = 0; i < 200_000; i += 1) crypto.createHash('sha256').update(sample).digest('hex');
const shaRate = 200_000 / ((performance.now() - start) / 1000);

const salt = bcrypt.genSaltSync(config.bcryptRounds);
start = performance.now();
for (let i = 0; i < 10; i += 1) bcrypt.hashSync(sample, salt);
const bcryptRate = 10 / ((performance.now() - start) / 1000);

const space = 26 ** 8; // every lower-case password of eight letters
const fixed0 = { maximumFractionDigits: 0 };
console.log(`SHA-256 : ${shaRate.toLocaleString('en-US', fixed0).padStart(15)} guesses/second, one CPU core`);
console.log(`bcrypt  : ${bcryptRate.toLocaleString('en-US', fixed0).padStart(15)} guesses/second, one CPU core`);
console.log(`ratio   : ${(shaRate / bcryptRate).toLocaleString('en-US', fixed0).padStart(15)}x`);
console.log(`\nA password of eight lower-case letters: ${space.toLocaleString('en-US')} possibilities.`);
console.log(`SHA-256 : ${(space / shaRate / 3600).toFixed(1)} hours on this one core`);
console.log(`bcrypt  : ${(space / bcryptRate / 31_557_600).toLocaleString('en-US', fixed0)} years on this one core`);
console.log('\nA real attacker owns a GPU farm, not one core of a laptop.');
console.log('The ratio is what matters, and the ratio does not change.');
